amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java

```diff
@@ -409,24 +390,42 @@ public void onConnectionStateChange(BluetoothGatt gatt, int status,
                                                     int newState) {
                     Log.i(TAG, "BLE connection state changed: " + status + "; new state: "
                             + AmazonFreeRTOSConstants.BleConnectionState.values()[newState]);
-                    String intentAction;
-                    if (newState == BluetoothProfile.STATE_CONNECTED) {
-                        //intentAction = ACTION_GATT_CONNECTED;
-                        mBleConnectionState = AmazonFreeRTOSConstants.BleConnectionState.BLE_CONNECTED;
-                        final IntentFilter bondFilter = new IntentFilter(BluetoothDevice.ACTION_BOND_STATE_CHANGED);
-                        mContext.registerReceiver(mBondStateBroadcastReceiver, bondFilter);
-                        //broadcastUpdate(intentAction);
-                        Log.i(TAG, "Connected to GATT server.");
-                        discoverServices();
-                        mBluetoothDevice.createBond();
-                    } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
-                        //intentAction = ACTION_GATT_DISCONNECTED;
+                    if (status == BluetoothGatt.GATT_SUCCESS) {
+                        if (newState == BluetoothProfile.STATE_CONNECTED) {
+                            int bondState = mBluetoothDevice.getBondState();
+
+                            mBleConnectionState = AmazonFreeRTOSConstants.BleConnectionState.BLE_CONNECTED;
+                            Log.i(TAG, "Connected to GATT server.");
+                            // If the device is already bonded or will not bond we can call discoverServices() immediately
+                            if (bondState == BluetoothDevice.BOND_NONE || bondState == BluetoothDevice.BOND_BONDED) {
+                                discoverServices();
+                            }
+                        } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
+                            //intentAction = ACTION_GATT_DISCONNECTED;
+                            mBleConnectionState = AmazonFreeRTOSConstants.BleConnectionState.BLE_DISCONNECTED;
+                            Log.i(TAG, "Disconnected from GATT server.");
+
+                            // If ble connection is closed, there's no need to keep mqtt connection open.
+                            //mBleConnectionState = AmazonFreeRTOSConstants.BleConnectionState.BLE_DISCONNECTED;
+                            if (mMqttConnectionState != AmazonFreeRTOSConstants.MqttConnectionState.MQTT_Disconnected) {
```

Comment:
We need to call this in  `else` block as well when there is a disconnect due to error.